### ISSUE TITLE
MONAD.ipynb doesn't replicate the paper

### ISSUE THREAD

**Issue by dmascali** [kevaldoshi17/Prediction-based-Video-Anomaly-Detection-#9]
Hey, I found your proposed method quite interesting. However, I am having issues in replicating the AUC reported in the paper (97.2) and in understanding the choice of some parameters in MONAD.ipynb. Could you please help me with the following points?

- Objects with confidence level above 0.97 (in the training set) and 0.6 (in the testing set) are used. However, in the paper the thresholds are 0.6 and 0.4, respectively. 
- The feature weighting parameters (i.e., w1, w2 and w3) are not exactly 1, 0.4 and 0.9 as reported in the paper. Indeed, while w1 is 1, w2 is set to 0.2 and w3 is set to 0 for the first class of yolo objects and set to 1 for the remaining classes. Moreover, in the Test set the yolo features which differ from zero are set to 0.9 (instead of being multiplied by 0.9).
- The error seems to be equal to the MSE plus the sum of the 10-NN distances of the remaining features. In the paper, it seems that the distance is calculated using all the features, including the MSE.  
- It seems that you are excluding the very last yolo class (see: X_N[i,1:-1]).  
- In your experience, the above parameters, including using k=10 in the NN distance, fit well with other datasets?
- For calculating the instantaneous frame-level anomaly evidence δt, you used a pre-computed, video-based, value derived from the base_lm variable (called in the code lm2). However, since I would like to use the method online, I am forced to calculate δt using base_lm only (i.e., using a global normalization). Are the AUC reported in the paper obtained with the global normalization (i.e., base_lm) or with the video-based normalization (i.e., lm2)?   

I tried switching back all the parameters to what reported in the paper, but I cannot get the AUC you reported. Could you please help me in replicating your results? 
Thank you very much!

**Comment by kevaldoshi17** [comment]
Hi, unfortunately I do not have access to my original code base at the moment so I will try to answer your questions as well as I can. 

1. Even after publishing the paper, we attempted to improve the results and hence were playing with the parameters to see if there was something else that might work better. By looking at the jupyter notebook, it looks like the current parameters gave a result of 97.7. 
2. If I remember it correctly, doing it either way did not matter much. By adding the MSE, we tried to reduce the dimensionality which affects the Knn computation. 
3. Thanks for pointing it out. That seems to be a mistake in the code, I'll fix it in the next commit. 
4. When we tested the parameters, generally the selection of k is not very sensitive, i.e., you can use a reasonably close value for k and expect good results on all datasets. 
5. The AUC reported in the paper is computed using the video-based normalization but we show that our method can also work online by using global normalization. 

Once I have access to my original codes I'll try to post them in a separate branch.

**Comment by Siftee150** [comment]
Hi, I also had a doubt. In the paper, it's written that mean, variance, skewness and kurtosis are extracted from the optical flow result. I searched a lot, I couldn't find where the extraction is taking place from the raw data. 
Thanks a lot.

**Comment by kevaldoshi17** [comment]
> Hi, I also had a doubt. In the paper, it's written that mean, variance, skewness and kurtosis are extracted from the optical flow result. I searched a lot, I couldn't find where the extraction is taking place from the raw data.
> Thanks a lot.

Hi, I think you might be referencing to a different paper here. In this work, we do not use optical flow or extract any features from it. Here's the link to this paper: https://arxiv.org/pdf/2010.07110.pdf

**Comment by Siftee150** [comment]
Oh yes very sorry, it's a different paper by you, this one: https://arxiv.org/abs/2004.07941
Is the repository for this available too?
Thanks a lot.

**Comment by dmascali** [comment]
> Hi, unfortunately I do not have access to my original code base at the moment so I will try to answer your questions as well as I can.
> 
> 1. Even after publishing the paper, we attempted to improve the results and hence were playing with the parameters to see if there was something else that might work better. By looking at the jupyter notebook, it looks like the current parameters gave a result of 97.7.
> 2. If I remember it correctly, doing it either way did not matter much. By adding the MSE, we tried to reduce the dimensionality which affects the Knn computation.
> 3. Thanks for pointing it out. That seems to be a mistake in the code, I'll fix it in the next commit.
> 4. When we tested the parameters, generally the selection of k is not very sensitive, i.e., you can use a reasonably close value for k and expect good results on all datasets.
> 5. The AUC reported in the paper is computed using the video-based normalization but we show that our method can also work online by using global normalization.
> 
> Once I have access to my original codes I'll try to post them in a separate branch.

Thank you for your kind reply.
Meanwhile, I found something that puzzles me. There seems to be a mismatch between the index of the class probabilities in the trainining FS and in the testing FS. In the training FS, the indexes are selected following the criterion "the first it comes". On the contrary, in the testing FS, the order of the classes are defined manually (e.g., Classes_Test = ['person', 'skateboard', 'bicycle', 'skis']).  When "person" is the only object found in the training set, this is not going to be a problem for the final inference. However, in more complex datasets I think it might be a problem. Thus, I was thinking of using the yolo's class index (I guess it goes from 0 to 80) to define the FS. This will dramatically increase the dimensionality of the FS, but, since many of the class will be zero, I don't think it's going to impact negatively on the inference. What do you think? Please, correct me if I am wrong or I am missing something. Thank you.
